app/master/cluster_master.py

```diff
@@ -62,12 +62,34 @@ def api_representation(self):
             'slaves': slaves_representation,
         }
 
-    def builds(self):
+    def builds(self, offset: int=None, limit: int=None) -> List['Build']:
         """
-        Returns a list of all builds
-        :rtype: list[Build]
+        Returns a list of all builds.
+        :param offset: The starting index of the requested build
+        :param limit: The number of builds requested
         """
-        return self._all_builds_by_id.values()
+        num_builds = len(self._all_builds_by_id)
+        highest_amount = min(num_builds, Configuration['pagination_max_limit'])
+
+        offset = offset if offset is not None else Configuration['pagination_offset']
+        limit = limit if limit is not None else Configuration['pagination_limit']
+
+        # Remove any negative values
+        offset = max(offset, 0)
+        limit = max(limit, 0)
+
+        # If limit is set higher than the number of builds/max limit, reduce limit
+        limit = highest_amount if limit > highest_amount else limit
+
+        # Requested offset too high should yield no results
+        if offset > num_builds:
+            return []
+
+        first_build_id = offset + 1
```

Comment:
I think using `islice` is extra dangerous now that you've changed the dict from an OrderedDict to a regular dict. `islice` returns things in iteration order, but iteration order is not defined for a dict. I expect that `islice` might return different results from machine to machine, or even from hour to hour on the same machine.

I say if you go back to `islice`, at least change the dict back to an OrderedDict. That way at least the iteration order is defined, and we can make guarantees about it. Why did you change it to a normal dict in the first place?